---
title: "RNASeq Comparison between Ubiquitous and iPSC Datasets"
author: Ellora Chua
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    theme: yeti
    highlight: haddock
---

```{r global_options, include=FALSE}
short=FALSE 
#if short==TRUE, do not echo code chunks -- code folding doesn't work if short == TRUE
debug=FALSE
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(fig.width=12, fig.height=8, fig.path='./figures/comparison/',
                      echo=!short, warning=debug, message=debug, dev=c("png", "cairo_pdf"))
```

```{r loading libraries, include = FALSE}
source("functions.R")
library(DESeq2)
library(ggplot2)
library(genefilter)
library(GenomicFeatures)
library(biomaRt)
library(knitr)
library(reshape2)
library(scales)
library(Biostrings)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(org.Dm.eg.db)
library(tidyverse)
library(pheatmap)
library(RColorBrewer)
library(VennDiagram)
library(DT)
library(enrichR)
library(ggpubr)
library(png)
library(clusterProfiler)
```

```{r, include=FALSE}
ensembl.genes <- readRDS("output/ubiquitous/ensembl.genes.rds")
ubiq_pam_clust <- readRDS("output/ubiquitous/pam_clust.rds")
ipsc_pam_clust <- readRDS("output/ipsc/pam_clust.rds")
if(!dir.exists("output/comparison/")){
  dir.create("output/comparison", recursive = TRUE)
}
```

# Heatmaps {.tabset}
## Ubiquitous
```{r, include = FALSE}
ubiq_pam_clust <- as.data.frame(ubiq_pam_clust)
ubiq_pam_clust$Cluster <- factor(ubiq_pam_clust$Cluster, levels = c(2,3,4,7,5,6,1))
ubiq_pam_clust <- ubiq_pam_clust[order(ubiq_pam_clust$Cluster),]
```

```{r}
experimental_metadata <- read.delim("data/ubiquitous/metadata_rem.txt", sep=",", header=TRUE, stringsAsFactors=FALSE)
annotation = data.frame(Condition = rep(c("Control","OKSM", "Senolytic", "Senolyic_OKSM"),
                                            c(3, 3, 2, 3)))

row.names(annotation) = experimental_metadata$sample_id 
anno_colours = list(Condition = c(Control = "#BF3100", OKSM = "#E9B44C", Senolytic = "#1B998B", Senolyic_OKSM = "#5D576B"))
```

```{r}
pheatmap(ubiq_pam_clust[,1:(ncol(ubiq_pam_clust)-1)],
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         fontsize_row = 5.5,
         annotation_col = annotation,
         annotation_colors = anno_colours,
         cluster_rows = FALSE,
         cluster_cols = FALSE)
```


```{r, include=FALSE}
pam_clust_df <- as.data.frame(ubiq_pam_clust)
pam_clust_df$gene_name <- ensembl.genes[rownames(pam_clust_df),]$external_gene_name
pam_clust1_genes <- pam_clust_df[pam_clust_df$Cluster == 2, ] %>%
  mutate(Cluster = 1)
pam_clust2_genes <- pam_clust_df[pam_clust_df$Cluster == 3, ] %>%
  mutate(Cluster = 2) 
pam_clust3_genes <- pam_clust_df[pam_clust_df$Cluster == 4, ] %>%
  mutate(Cluster = 3)
pam_clust4_genes <- pam_clust_df[pam_clust_df$Cluster == 7, ] %>%
  mutate(Cluster = 4)
pam_clust5_genes <- pam_clust_df[pam_clust_df$Cluster == 5, ]
pam_clust6_genes <- pam_clust_df[pam_clust_df$Cluster == 6, ]
pam_clust7_genes <- pam_clust_df[pam_clust_df$Cluster == 1, ] %>%
  mutate(Cluster = 7)
pam_clust_df = rbind(pam_clust1_genes, pam_clust2_genes, pam_clust3_genes, pam_clust4_genes, pam_clust5_genes, pam_clust6_genes,pam_clust7_genes)
```

```{r}
pam_clust_df %>% 
  dplyr::select(gene_name, Cluster) %>%
  datatable(options = list(scrollX = TRUE), class = "white-space: nowrap")
```

```{r}
ubi_c1 <- pam_clust_df[pam_clust_df$Cluster == 1, ] %>%
  dplyr::select(-Cluster)

ubi_c2 <- pam_clust_df[pam_clust_df$Cluster == 2, ] %>%
  dplyr::select(-Cluster)

ubi_c3 <- pam_clust_df[pam_clust_df$Cluster == 3, ] %>%
  dplyr::select(-Cluster)

ubi_c4 <- pam_clust_df[pam_clust_df$Cluster == 4, ] %>%
  dplyr::select(-Cluster)

ubi_c5 <- pam_clust_df[pam_clust_df$Cluster == 5, ] %>%
  dplyr::select(-Cluster)

ubi_c6 <- pam_clust_df[pam_clust_df$Cluster == 6, ] %>%
  dplyr::select(-Cluster)

ubi_c7 <- pam_clust_df[pam_clust_df$Cluster == 7, ] %>%
  dplyr::select(-Cluster)

data.frame(Cluster = c("Cluster 1", "Cluster 2", "Cluster 3", "Cluster 4", "Cluster 5", 
                       "Cluster 6", "Cluster 7", "Total"),
                        Number_of_genes = c(nrow(ubi_c1), nrow(ubi_c2), nrow(ubi_c3), nrow(ubi_c4),nrow(ubi_c5), 
                                            nrow(ubi_c6), nrow(ubi_c7),
                                            sum(c(nrow(ubi_c1),nrow(ubi_c2),nrow(ubi_c3),nrow(ubi_c4),nrow(ubi_c5),
                                                  nrow(ubi_c6),nrow(ubi_c7)))))
```

## iPSC
```{r, include=FALSE}
experimental_metadata <- read.delim("data/ipsc/metadata.txt", sep=",", header=TRUE, stringsAsFactors=FALSE)
annotation = data.frame(Condition = rep(c("Control","OKSM", "Senolytic", "Senolyic_OKSM"),
                                            c(3, 3, 3, 3)))

row.names(annotation) = experimental_metadata$sample_id 
anno_colours = list(Condition = c(Control = "#BF3100", OKSM = "#E9B44C", Senolytic = "#1B998B", Senolyic_OKSM = "#5D576B"))
```

```{r}
pheatmap(ipsc_pam_clust[,1:(ncol(ipsc_pam_clust)-1)],
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         fontsize_row = 5.5,
         annotation_col = annotation,
         annotation_colors = anno_colours,
         cluster_rows = FALSE,
         cluster_cols = FALSE)

```

```{r}
ipsc_pam_clust_df <- ipsc_pam_clust %>%
  as.data.frame() %>%
  mutate(gene_name = ensembl.genes[rownames(.),]$external_gene_name)
datatable(ipsc_pam_clust_df, options = list(scrollX = TRUE), class = "white-space: nowrap")
```

```{r}
ipsc_c1 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 1, ] %>%
  dplyr::select(-Cluster)

ipsc_c2 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 2, ] %>%
  dplyr::select(-Cluster)

ipsc_c3 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 3, ] %>%
  dplyr::select(-Cluster)

ipsc_c4 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 4, ] %>%
  dplyr::select(-Cluster)

ipsc_c5 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 5, ] %>%
  dplyr::select(-Cluster)

ipsc_c6 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 6, ] %>%
  dplyr::select(-Cluster)

ipsc_c7 <- ipsc_pam_clust_df[ipsc_pam_clust_df$Cluster == 7, ] %>%
  dplyr::select(-Cluster)

data.frame(Cluster = c("Cluster 1", "Cluster 2", "Cluster 3", "Cluster 4", 
                       "Cluster 5", "Cluster 6", "Cluster 7", "Total"),
                        Number_of_genes = c(nrow(ipsc_c1), nrow(ipsc_c2), nrow(ipsc_c3), nrow(ipsc_c4),
                                            nrow(ipsc_c5), nrow(ipsc_c6), nrow(ipsc_c7),
                                            sum(c(nrow(ipsc_c1),nrow(ipsc_c2),nrow(ipsc_c3),nrow(ipsc_c4),
                                                  nrow(ipsc_c5),nrow(ipsc_c6),nrow(ipsc_c7)))))
```


# Gene Expression Boxplots {.tabset}
## Ubiquitous {.tabset}
### Cluster 1
```{r}
cluster_boxplot(ubi_c1, "Cluster 1", "TdTom", "Sen_OKSM", "OKSM")
```

### Cluster 2
```{r}
cluster_boxplot(ubi_c2, "Cluster 2", "TdTom", "Sen_OKSM", "OKSM")
```

### Cluster 3
```{r}
cluster_boxplot(ubi_c3, "Cluster 3", "TdTom", "Sen_OKSM", "OKSM")
```

### Cluster 4
```{r}
cluster_boxplot(ubi_c4, "Cluster 4", "TdTom", "Sen_OKSM", "OKSM")
```

### Cluster 5
```{r}
cluster_boxplot(ubi_c5, "Cluster 5", "TdTom", "Sen_OKSM", "OKSM")
```

### Cluster 6
```{r}
cluster_boxplot(ubi_c6, "Cluster 6", "TdTom", "Sen_OKSM", "OKSM")
```

### Cluster 7
```{r}
cluster_boxplot(ubi_c7, "Cluster 7", "TdTom", "Sen_OKSM", "OKSM")
```

## iPSC {.tabset}
### Cluster 1
```{r}
cluster_boxplot(ipsc_c1, "Cluster 1", "Control", "SO", "O")
```

### Cluster 2
```{r}
cluster_boxplot(ipsc_c2, "Cluster 2", "Control", "SO", "O")
```

### Cluster 3
```{r}
cluster_boxplot(ipsc_c3, "Cluster 3", "Control", "SO", "O")
```

### Cluster 4
```{r}
cluster_boxplot(ipsc_c4, "Cluster 4", "Control", "SO", "O")
```

### Cluster 5
```{r}
cluster_boxplot(ipsc_c5, "Cluster 5", "Control", "SO", "O")
```

### Cluster 6
```{r}
cluster_boxplot(ipsc_c6, "Cluster 6", "Control", "SO", "O")
```

### Cluster 7
```{r}
cluster_boxplot(ipsc_c7, "Cluster 7", "Control", "SO", "O")
```

#  Gene overlaps
## Overall
```{r, include=FALSE}
venn.diagram(x=list(Ubiquitous = na.omit(rownames(pam_clust_df)), iPSC = na.omit(rownames(ipsc_pam_clust))), fill = c("#ffc93c", "#f8a488"), margin = 0.08, cat.pos = c(90,270), cat.dist = c(0.03,0.05), alpha = c(0.7, 0.7), imagetype = "png", print.mode = "raw", filename = "output/comparison/overlaps.png")
```

```{r}
olaps <- readPNG("output/comparison/overlaps.png")
p <- rasterGrob(olaps, interpolate = TRUE)
ggarrange(p)
```

I've manipulated the heatmaps such that Clusters 1 - 5 in the ubiquitous data is the same as Clusters 1 - 7 in the iPSC dataset for easy comparison

## C1 ubiquitous dataset in C1 and/or C2 of iPSC dataset
Percentage of genes in c1 of ubiquitous dataset in c1 & c2 of iPSC dataset
```{r}
denom <- c(rownames(ipsc_c1),rownames(ipsc_c2))
sum(rownames(ubi_c1) %in% denom)/length(denom)*100
```

Where else can we find c1 ubiquitous genes in iPSC dataset?
```{r}
tab <- ipsc_pam_clust_df[rownames(ubi_c1),] %>%
  drop_na %>%
  group_by(Cluster) %>%
  tally()
tab
```

`r sum(tab$n)` genes in Cluster 1 of the ubiquitous dataset is accounted for in the iPSC dataset. This implies that `r length(denom)-sum(tab$n)` genes in C1 of the ubiquitous dataset is not expressed or significantly differentially expressed in the iPSC dataset. 

## C3 ubiquitous dataset in C4 and or C5 of iPSC dataset
Percentage of genes in c3 of ubiquitous dataset in c4 & c5 iPSC dataset
```{r}
denom <- c(rownames(ipsc_c4),rownames(ipsc_c5))
sum(rownames(ubi_c3) %in% denom)/length(denom)*100
```

Where else can we find c3 ubiquitous genes in iPSC dataset?
```{r}
tab <- ipsc_pam_clust_df[rownames(ubi_c3),] %>%
  drop_na %>%
  group_by(Cluster) %>%
  tally()
tab
```

`r sum(tab$n)` genes in Cluster 3 of the ubiquitous dataset is accounted for in the iPSC dataset. This implies that `r length(denom)-sum(tab$n)` genes in C3 of the ubiquitous dataset is not expressed or significantly differentially expressed in the iPSC dataset. 

## C4 ubiquitous dataset in C6 of iPSC dataset
Percentage of genes in c4 of ubiquitous dataset in c6 iPSC dataset
```{r}
sum(rownames(ubi_c4) %in% rownames(ipsc_c6))/length(rownames(ipsc_c6))*100
```

Where else can we find c4 ubiquitous genes in iPSC dataset?
```{r}
tab <- ipsc_pam_clust_df[rownames(ubi_c4),] %>%
  drop_na %>%
  group_by(Cluster) %>%
  tally()
tab
```

`r sum(tab$n)` genes in Cluster 4 of the ubiquitous dataset is accounted for in the iPSC dataset. This implies that `r length(rownames(ipsc_c6))-sum(tab$n)` genes in C4 of the ubiquitous dataset is not expressed or significantly differentially expressed in the iPSC dataset. 


# Summary Enrichments from compareCluster {.tabset}
```{r, include=FALSE}
library(ReactomePA)
ddssva = readRDS("output/ubiquitous/ddssva.rds")
```

## Ubiquitous
```{r}
genes = lapply(list(c1=ubi_c1, c2=ubi_c2, c3=ubi_c3, c4=ubi_c4, c5=ubi_c5, c6=ubi_c6, c7=ubi_c7), function(i){ensembl.genes[rownames(i)]$entrezgene_id})
bg = ensembl.genes[rownames(ddssva)]$entrezgene_id
```

```{r}
compPathway <- compareCluster(geneCluster   = genes,
                              fun           = "enrichGO",
                              keyType       = "ENTREZID",
                              universe      = bg,
                              ont           = "BP",
                              OrgDb         = "org.Dm.eg.db",
                              pvalueCutoff  = 1,
                              qvalueCutoff  = 0.1,
                              pAdjustMethod = "BH")
dotplot(compPathway, showCategory = 10, title = "GO Enrichment Analysis") + theme(axis.text.y = element_text(size=8))
```

```{r}
compPathway <- compareCluster(geneCluster   = genes,
                              fun           = "enrichKEGG",
                              keyType       = "ncbi-geneid",
                              universe      = bg,
                              organism      = "dme",
                              pvalueCutoff  = 1,
                              qvalueCutoff  = 0.2,
                              pAdjustMethod = "BH")
dotplot(compPathway, showCategory = 10, title = "KEGG Enrichment Analysis") + theme(axis.text.y = element_text(size=8))
```


## iPSC
```{r}
genes = lapply(list(c1=ipsc_c1, c2=ipsc_c2, c3=ipsc_c3, c4=ipsc_c4, c5=ipsc_c5, c6=ipsc_c6, c7=ipsc_c7), function(i){ensembl.genes[rownames(i)]$entrezgene_id})
```

```{r}
compPathway <- compareCluster(geneCluster   = genes,
                              fun           = "enrichGO",
                              keyType       = "ENTREZID",
                              ont           = "BP",
                              OrgDb         = "org.Dm.eg.db",
                              pvalueCutoff  = 1,
                              qvalueCutoff  = 0.1,
                              pAdjustMethod = "BH")
dotplot(compPathway, showCategory = 8, title = "GO Enrichment Analysis") + theme(axis.text.y = element_text(size=8))
```

```{r}
compPathway <- compareCluster(geneCluster   = genes,
                              fun           = "enrichKEGG",
                              keyType       = "ncbi-geneid",
                              organism      = "dme",
                              pvalueCutoff  = 1,
                              qvalueCutoff  = 0.1,
                              pAdjustMethod = "BH")
dotplot(compPathway, showCategory = 10, title = "KEGG Enrichment Analysis") + theme(axis.text.y = element_text(size=8))
```

# Summary Enrichments from custom script {.tabset}
## Ubiquitous
```{r}
filenames = list.files("output/ubiquitous", pattern = "_go.rds", full.names = TRUE)
all_go = lapply(filenames, function(x){readRDS(x)})
names(all_go) <- c("c1", "c2", "c3", "c4", "c5", "c6", "c7")
ubi_go <- all_go %>%
  bind_rows(.id = "Cluster") %>%
  group_by(Cluster) %>%
  slice_min(order_by = Adjusted.P.value, n = 10)
```

```{r}
ubi_go %>%
  mutate(Term = gsub("\\([^()]*\\)", "", Term),
         Term = str_to_title(Term)) %>%
  mutate(Cluster = factor(Cluster, levels = c("c1","c2", "c3", "c4","c5","c6","c7"))) %>%
  group_by(Cluster) %>%
  mutate(Term = factor(Term, levels = Term)) %>%
  ggplot(aes(x = Cluster, y = Term)) +
  geom_point(aes(colour = Adjusted.P.value, size = Ratio)) + 
  ylab(NULL) + 
  xlab("Cluster") +
  scale_color_continuous(low="red", high="blue", guide=guide_colorbar(reverse=FALSE)) + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 60),
                   limits=rev) +
  #labs(fill = "Adjusted p-value") +
  guides(colour = guide_colorbar(title = "Adjusted p-value", reverse = TRUE),
         size = guide_legend(title = "Gene Ratio")) +
  theme(axis.text.x = element_text(size=8)) +
  theme_bw() +
  ggtitle("GO Enrichment Analysis")

```

```{r}
filenames = list.files("output/ubiquitous", pattern = "_kegg.rds", full.names = TRUE)
all_kegg = lapply(filenames, function(x){readRDS(x)})
names(all_kegg) <- c("c1", "c2", "c4", "c5", "c6", "c7")
ubi_kegg <- all_kegg %>%
  bind_rows(.id = "Cluster") %>%
  group_by(Cluster) %>%
  slice_min(order_by = Adjusted.P.value, n = 10)
```

```{r}
ubi_kegg %>%
  mutate(Term = gsub("\\([^()]*\\)", "", Term),
         Term = str_to_title(Term)) %>%
  mutate(Cluster = factor(Cluster, levels = c("c1","c2","c4", "c5","c6","c7"))) %>%
  group_by(Cluster) %>%
  mutate(Term = factor(Term, levels = Term)) %>%
  ggplot(aes(x = Cluster, y = Term)) +
  geom_point(aes(colour = Adjusted.P.value, size = Ratio)) + 
  ylab(NULL) + 
  xlab("Cluster") +
  scale_color_continuous(low="red", high="blue", guide=guide_colorbar(reverse=FALSE)) + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 60),
                   limits=rev) +
  #labs(fill = "Adjusted p-value") +
  guides(colour = guide_colorbar(title = "Adjusted p-value", reverse = TRUE),
         size = guide_legend(title = "Gene Ratio")) +
  theme(axis.text.x = element_text(size=8)) +
  theme_bw() +
  ggtitle("KEGG Enrichment Analysis")

```

# Session Info
```{r}
sessionInfo()
```