---
title: "OKSM_Senolytic RNA-Seq Analysis (Pairwise)"
author: Ellora Chua
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    theme: yeti
    highlight: haddock
---

```{r global_options, include=FALSE}
short=FALSE 
#if short==TRUE, do not echo code chunks -- code folding doesn't work if short == TRUE
debug=FALSE
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(fig.width=12, fig.height=8, fig.path='./figures/ubiquitous/pairwise/',
                      echo=!short, warning=debug, message=debug, dev=c("png", "cairo_pdf"))
```

```{r loading libraries, include = FALSE}
source("functions.R")
library(DESeq2)
library(ggplot2)
library(genefilter)
library(GenomicFeatures)
library(biomaRt)
library(knitr)
library(reshape2)
library(scales)
library(Biostrings)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(tidyverse)
library(pheatmap)
library(RColorBrewer)
library(cowplot)
library(DT)
library(enrichR)
```

# Quality Control & EDA
```{r cache=TRUE, include=FALSE}
dm.gtf.db <- makeTxDbFromGFF("annotation/Drosophila_melanogaster.BDGP6.22.97.chr.gtf", format="gtf" )
ensembl.genes = genes(dm.gtf.db)
fruitfly = useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", dataset="dmelanogaster_gene_ensembl", mirror = "uswest")
bm.annotations = getBM(attributes=c("ensembl_gene_id", "entrezgene_id", "gene_biotype", "external_gene_name"), mart=fruitfly, filters="ensembl_gene_id", values=ensembl.genes$gene_id, uniqueRows=TRUE)
ensembl.genes$gene_biotype = bm.annotations$gene_biotype[match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ensembl.genes$entrezgene_id = bm.annotations$entrezgene_id[match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ensembl.genes$external_gene_name = bm.annotations$external_gene_name[match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
```

```{r loading data, include=FALSE}
experimental_metadata = read.delim("data/ubiquitous/pairwise.txt", sep=",", header=TRUE, stringsAsFactors=FALSE)

#Create a matrix of the data (nrow can be obtained by calling wc -l *.genes.results in terminal; note -1 for header)
#Note that the nrow for this now changes since we've added the construct
data = matrix(0, ncol=length(experimental_metadata$sample_id), nrow=17714)
colnames(data)= experimental_metadata$sample_id
for( i in experimental_metadata$sample_id){
  data[,i] = read.table(paste("data/ubiquitous/", i, ".genes.results",sep=""), sep="\t", header=TRUE)$expected_count
}
row.names(data) = read.table(paste("data/ubiquitous/", i, ".genes.results",sep=""), sep="\t", header=TRUE)$gene_id

#Create a factor for the condition column - by making it a factor you give it an order
experimental_metadata$condition = factor(experimental_metadata$condition, levels=c("Control", "OKSM"))

#Create a factor for the replicate column - by making it a factor you give it an order
experimental_metadata$replicate = factor(experimental_metadata$replicate, levels=c("1", "2", "3"))
#experimental_metadata$batch = factor(experimental_metadata$batch, levels=c("1", "2", "3"))

#DESeq needs read counts to be integers
data_mat = apply(round(data), c(1,2), as.integer)

#Specifying which group is the control group to make appropriate contrasts downstream
#experimental_metadata$condition <- relevel(experimental_metadata$condition, ref = "Control")

#Creates the DESeq2 object
#Since there aren't any batch effects, we don't need to take those into account
dds = DESeqDataSetFromMatrix(data_mat, experimental_metadata, ~ condition)
colnames(dds) = colnames(data_mat)

#Estimates the size factors using the "median ratio method" described by Equation 5 in Anders and Huber (2010)
dds <- estimateSizeFactors(dds) 

#This function obtains dispersion estimates for a count data set
dds <- estimateDispersions(dds)

#This function transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts
rld <- rlog(dds)
```

## All samples (Before SVA) {.tabset}
### Sample clustering
```{r hclust}
sampleDists <- dist(t(assay(rld)))
plot(hclust(sampleDists))
```

### Correlation Heatmap
```{r}
annot = dplyr::select(experimental_metadata, condition)
row.names(annot) = experimental_metadata$sample_id
rld %>%
  assay() %>%
  cor() %>%
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
           annotation = annot,
           annotation_colors = list(condition = c(Control = "#BF3100", OKSM = "#E9B44C", 
                                                  Senolytic = "#1B998B", Senolytic_OKSM = "#5D576B")),
           cluster_rows = TRUE,
           cluster_cols = T,
           cellwidth = 13,
           cellheight = 13)
```

### PCA
```{r pca_samples, results='hide'}
ntop = 500
rv <- rowVars(assay(rld))
select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
pca = prcomp(t(assay(rld)[select,]))
percentVar <- pca$sdev^2/sum(pca$sdev^2)

pca_data <- plotPCA(rld, intgroup = c("condition", "replicate"), returnData=TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"), digits=2)
ggplot(pca_data, aes(PC1, PC2, color=condition, shape=replicate)) + geom_point(size=3) +
  scale_x_continuous(paste0("PC1: ",percentVar[1],"% variance")) +
  scale_y_continuous(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + theme_classic() + geom_text(data = pca_data, aes(PC1,PC2, label = name), hjust = 1.2)

```

## SVA {.tabset}

From the hierarchical clustering, correlation heatmap, and PCA, there seems to be some variation that we're not accounting for. Here, we use SVA to account for the unknown variation. Since the number of variables is unknown, we call the sva function without the `n.sv` argument, allowing the algorithm to estimate the number of factors.

```{r}
# There are batch effects for this dataset. However, running ~ batch + condition gives us an error that says the model matrix is not full rank. So we follow this guide https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#removing-hidden-batch-effects
library(sva)
dat  <- counts(dds, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~ condition, colData(rld))
mod0 <- model.matrix(~ 1, colData(rld))
svseq <- svaseq(dat, mod, mod0)
```

According to SVA, there are 2 significant surrogate variables. 

```{r}
par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(svseq$sv[, i] ~ dds$condition, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
 }
```

```{r, include=FALSE}
ddssva <- dds
ddssva$SV1 <- svseq$sv[,1]
ddssva$SV2 <- svseq$sv[,2]
design(ddssva) <- ~ SV1 + SV2 + condition

#Estimates the size factors using the "median ratio method" described by Equation 5 in Anders and Huber (2010)
ddssva <- estimateSizeFactors(ddssva)

#This function obtains dispersion estimates for a count data set
ddssva <- estimateDispersions(ddssva)

#This function transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts
rld <- rlog(ddssva)
```

After accounting for the 2 significant surrogate variables, this is what the data looks like:

### Sample clustering
```{r, include=FALSE}
cov = cbind(rld$SV1,rld$SV2)
assay(rld) <- limma::removeBatchEffect(assay(rld), covariates = cov)
```

```{r}
sampleDists <- dist(t(assay(rld)))
plot(hclust(sampleDists))
```

### Correlation
```{r, corr_sva}
rld %>%
  assay() %>%
  cor() %>%
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
           annotation = annot,
           annotation_colors = list(condition = c(Control = "#BF3100", OKSM = "#E9B44C", 
                                                  Senolytic = "#1B998B", Senolytic_OKSM = "#5D576B")),
           cluster_rows = TRUE,
           cluster_cols = T,
           cellwidth = 13,
           cellheight = 13)
```

### PCA
```{r, pca_sva, results='hide'}
ntop = 500
rv <- rowVars(assay(rld))
select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
pca = prcomp(t(assay(rld)[select,]))
percentVar <- pca$sdev^2/sum(pca$sdev^2)

pca_data <- plotPCA(rld, intgroup = c("condition", "replicate"), returnData=TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"), digits=2)
ggplot(pca_data, aes(PC1, PC2, color=condition, shape=replicate)) + geom_point(size=3) +
  scale_x_continuous(paste0("PC1: ",percentVar[1],"% variance")) +
  scale_y_continuous(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed() + theme_classic() + geom_text(data = pca_data, aes(PC1,PC2, label = name), hjust = 1.2)
```


```{r remove rRNA, include=FALSE}
ddssva = ddssva[!(row.names(counts(ddssva)) %in% ensembl.genes$gene_id[ensembl.genes$gene_biotype %in% c("rRNA", "snoRNA", "snRNA")]),]
ddssva = ddssva[rowSums(counts(ddssva)) > 0,]
```

## Removing TPM Outliers
```{r tpm}
effective_lengths = matrix(0, ncol=length(experimental_metadata$sample_id), nrow=17714)
colnames(effective_lengths)= experimental_metadata$sample_id
for( i in experimental_metadata$sample_id){
  effective_lengths[,i] = read.table(paste("data/ubiquitous/", i, ".genes.results",sep=""), sep="\t", header=TRUE)$effective_length
}
row.names(effective_lengths) = read.table(paste("data/ubiquitous/", i, ".genes.results",sep=""), sep="\t", header=TRUE)$gene_id

effective_lengths = rowMeans(effective_lengths[row.names(counts(ddssva)),])
ncrpk = counts(ddssva) / (effective_lengths / 1000)
ncrpk = apply(ncrpk, c(1,2), function(x){if(is.nan(x)){0}else{x}})
ncrpk = apply(ncrpk, c(1,2), function(x){if(is.infinite(x)){0}else{x}})
ncscalingfactor = colSums(ncrpk) / 1e6
nctpm = sweep(ncrpk, 2, ncscalingfactor, "/")

nctpm.melt = melt(nctpm)
ggplot(nctpm.melt, aes(x=Var2, y=value)) + geom_boxplot() + theme_classic() + theme(axis.text.x = element_text(angle = 90, colour="black", hjust = 1)) + scale_x_discrete("Sample") + scale_y_continuous("TPM")
```

```{r}
tpm.threshold = 10000
test.tpm = apply(nctpm, 1, function(x){ any(x> tpm.threshold) })
ensembl.genes[names(test.tpm[test.tpm])] %>%
  as.data.frame() %>%
  datatable(options = list(scrollX = TRUE))
```
```{r, include=FALSE}
ddssva <- ddssva[!(row.names(counts(ddssva)) %in% ensembl.genes[names(test.tpm[test.tpm])]$gene_id),]
ddssva <- estimateSizeFactors(ddssva)
ddssva <- estimateDispersions(ddssva)
rld <- rlog(ddssva)
```

# Differential Expression Analysis
```{r, include=FALSE}
## Removing lowly expressed genes, only to be done once at the start of the differential expression step
filter = apply(counts(ddssva, normalized=TRUE), 1, function(x){ mean(x) >= 10 })
ddssva = ddssva[filter, ]
```

## Wald Tests {.tabset}
```{r, include=FALSE}
dds_wald = nbinomWaldTest(ddssva)
lfc.threshold = log2(1.5)
if(!dir.exists("output/ubiquitous/pairwise")){
  dir.create(file.path(getwd(),"output/ubiquitous/pairwise"), recursive = TRUE)
}
output_dir <- "output/ubiquitous/pairwise/"
```

### OKSM vs Control(TdTom)
```{r}
res1 <- results(dds_wald, contrast = c( "condition", "OKSM", "Control"), test="Wald")
res <- lfcShrink(dds = dds_wald, res = res1, type = "normal", contrast = c( "condition", "OKSM", "Control"))
res$gene_biotype <- ensembl.genes$gene_biotype[match(row.names(res1), ensembl.genes$gene_id)]
res$external_gene_name <- ensembl.genes$external_gene_name[match(row.names(res1), ensembl.genes$gene_id)]
print(paste("Number of significant genes (padj < 0.1 & log2FoldChange < log2(1.5)):", sum(res$padj < 0.1 & abs(res$log2FoldChange) > lfc.threshold, na.rm = T)))
hist(res$pvalue)
write_files(res, "OKSM", "Control")
EnhancedVolcano(res, lab=res$external_gene_name, pCutoff = 0.05, FCcutoff = lfc.threshold,
                  x = "log2FoldChange", y = "padj",
                  title = NULL,
                  subtitle = NULL,
                  legendPosition = 'right',
                  legendLabSize = 8)
```


```{r extract significant transcripts, include=FALSE}
## Get significant genes by 10% FDR & FC
sig_padj_genes <- res[!is.na(res$padj),]
sig_padj_genes <- sig_padj_genes[sig_padj_genes$padj < 0.1,]
sig_padj_fc_genes <- sig_padj_genes[abs(sig_padj_genes$log2FoldChange) > lfc.threshold,]

## Get the rld for all differentially expressed genes & rld for significant genes
rld <- rlog(ddssva)
significant_rld <- rld[rownames(sig_padj_fc_genes)]
```

```{r for getting granges for motif enrichment analysis, include=FALSE}
assay_df = assay(significant_rld)

## Differentially expressed genes
sig_de_granges <- ensembl.genes[rownames(assay_df)]
sig_de_granges <- promoters(sig_de_granges, upstream= 500, downstream = 500, use.names = TRUE)

# ## All expressed genes
# sig_all_granges <- ensembl.genes[rownames(counts(dds)),]
# sig_all_granges <- promoters(sig_all_granges, upstream = 500, downstream = 500, use.names = TRUE)
```

# Visualisation
```{r z-score transformation}
# assay(x) to access the count data
assay(significant_rld) <- limma::removeBatchEffect(assay(significant_rld), covariates = cov)
sig_mat_rld = assay(significant_rld)

# The apply function swaps the rows to samples and columns to genes -- the standard is the other way around: samples in cols and genes in rows, hence the transpose function
zscores = t(apply(sig_mat_rld, 1, function(x){ (x - mean(x)) / sd(x) }))
```

```{r heatmap annotation, include=FALSE}
## >> !! NOTE !! The annotation needs to be in the correct order as the condition in the metadata.txt
annotation = data.frame(Condition = rep(c("Control","OKSM"),
                                            c(3, 3)))

row.names(annotation) = experimental_metadata$sample_id 
anno_colours = list(Condition = c(Control = "#BF3100", OKSM = "#E9B44C"))
```

## Heatmap
```{r}
pam_clust <- generate_data(zscores, 2, "pam")
# pam_clust <- as.data.frame(pam_clust)
# pam_clust$Cluster <- factor(pam_clust$Cluster, levels = c(5,4,3,1,2,6))
# pam_clust <- pam_clust[order(pam_clust$Cluster),]

pheatmap(pam_clust[,1:(ncol(pam_clust)-1)],
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         fontsize_row = 5.5,
         annotation_col = annotation,
         annotation_colors = anno_colours,
         cluster_rows = FALSE,
         cluster_cols = FALSE)
```

## Table of Genes
```{r}
pam_clust_df <- pam_clust %>%
  as.data.frame() %>%
  mutate(gene_name = ensembl.genes[rownames(.),]$external_gene_name) %>%
  dplyr::select(gene_name, Cluster) %>%
  dplyr::rename("Gene Name" = gene_name)
datatable(pam_clust_df, options = list(scrollX = TRUE), class = "white-space: nowrap")
```

```{r, include = FALSE}
dds_df <- res %>%
  as.data.frame() %>%
  mutate(Cluster = pam_clust_df$Cluster[match(rownames(res), rownames(pam_clust_df))]) %>%
  dplyr::select(pvalue, padj, gene_biotype, external_gene_name, Cluster)
write.table(dds_df, "output/ubiquitous/pairwise/clusters.csv", row.names = TRUE, col.names = TRUE)
```

## Number of Genes
```{r}
c1 <- pam_clust_df[pam_clust_df$Cluster == 1, ] %>%
  dplyr::select(-Cluster)

c2 <- pam_clust_df[pam_clust_df$Cluster == 2, ] %>%
  dplyr::select(-Cluster)

data.frame(Cluster = c("Cluster 1", "Cluster 2", "Total"),
                        Number_of_genes = c(nrow(c1), nrow(c2),
                                            sum(c(nrow(c1),nrow(c2)))))
```

## Silhouette Plot
```{r}
## Checking the stability of clusters
#fviz_nbclust(zscores, kmeans, method = "silhouette")
set.seed(2)
dd = as.dist((1 - cor(t(zscores)))/2)
pam = pam(dd, 2, diss = TRUE)
sil = silhouette(pam$clustering, dd)
plot(sil, border=NA, main = "Silhouette Plot for 2 Clusters")
```

# Functional Enrichments
```{r}
#listEnrichrSites()
setEnrichrSite("FlyEnrichr")
dbs <- listEnrichrDbs()
to_check <- c("GO_Biological_Process_2018", "KEGG_2019")
```

## Cluster 1 {.tabset}
```{r, include=FALSE}
c1_genename = c1$`Gene Name`
eresList <- enrichr(c1_genename, databases = to_check)
```

### GO_Biological_Process_2018
```{r c1_go_7, fig.width=8}
eresList$GO_Biological_Process_2018 %>%
  plot_enrichr("GO_Biological_Process_2018")
datatable(eresList[[1]], options = list(scrollX = TRUE), class = "white-space: nowrap")
```

### KEGG_2019
```{r c1_kegg_7, fig.width=8}
eresList$KEGG_2019 %>%
  plot_enrichr("KEGG_2019") 
datatable(eresList[[2]], options = list(scrollX = TRUE), class = "white-space: nowrap")
```

## Cluster 2 {.tabset}
```{r, include=FALSE}
c2_genename = c2$`Gene Name`
eresList <- enrichr(c2_genename, databases = to_check)
```

### GO_Biological_Process_2018
```{r c2_go_7, fig.width=8}
eresList$GO_Biological_Process_2018 %>%
  plot_enrichr("GO_Biological_Process_2018")
datatable(eresList[[1]], options = list(scrollX = TRUE), class = "white-space: nowrap")
```

### KEGG_2019
```{r c2_kegg_7, fig.width=8}
eresList$KEGG_2019 %>%
  plot_enrichr("KEGG_2019") 
datatable(eresList[[2]], options = list(scrollX = TRUE), class = "white-space: nowrap")
```

# Motif Enrichment Analysis {.tabset}
```{r data preprocessing, include=FALSE}
## Output Dir
if(!dir.exists("ame/pairwise/input")){
  dir.create("ame/pairwise/input", recursive = TRUE)
}

output_dir <- "ame/pairwise/input/"

## Cluster 1
generate_fasta(output_dir, sig_de_granges, c1, "c1")
extract_control_deg(output_dir, sig_de_granges, c1, "c1_deg")

### Cluster 2
generate_fasta(output_dir, sig_de_granges, c2,"c2")
extract_control_deg(output_dir, sig_de_granges, c2, "c2_deg")
```

## Cluster 1
```{r}
c1_ame <- read.delim("ame/pairwise/output/c1/0.3/ame.tsv")
c1_ame %>%
  dplyr::select(-motif_DB, -motif_ID) %>%
  datatable(options = list(scrollX = TRUE))
```

## Cluster 2
```{r}
c2_ame <- read.delim("ame/pairwise/output/c2/0.3/ame.tsv")
c2_ame %>%
  dplyr::select(-motif_DB, -motif_ID) %>%
  datatable(options = list(scrollX = TRUE))
```

# Session Info
```{r}
sessionInfo()
```